pinot-connectors/presto-pinot-driver/src/main/java/org/apache/pinot/connector/presto/PinotScatterGatherQueryClient.java

```diff
@@ -0,0 +1,340 @@
+      String pql,
+      String serverHost,
+      List<String> segments,
+      long connectionTimeoutInMillis,
+      boolean ignoreEmptyResponses,
+      int pinotRetryCount) {
+    BrokerRequest brokerRequest;
+    try {
+      brokerRequest = REQUEST_COMPILER.compileToBrokerRequest(pql);
+    } catch (Pql2CompilationException e) {
+      throw new PinotException(ErrorCode.PINOT_INVALID_PQL_GENERATED,
+          String.format("Parsing error with on %s, Error = %s", serverHost, e.getMessage()), e);
+    }
+
+    Map<org.apache.pinot.core.transport.ServerInstance, List<String>> routingTable = new HashMap<>();
+    routingTable.put(new org.apache.pinot.core.transport.ServerInstance(new InstanceConfig(serverHost)),
+        new ArrayList<>(segments));
+
+    // Unfortunately the retries will all hit the same server because the routing decision has already been made by
+      // the pinot broker
+    Map<ServerInstance, DataTable> serverResponseMap = doWithRetries(pinotRetryCount, (requestId) -> {
+      String rawTableName = TableNameBuilder.extractRawTableName(brokerRequest.getQuerySource().getTableName());
+      if (!_concurrentQueriesCountMap.containsKey(serverHost)) {
+        _concurrentQueriesCountMap.put(serverHost, new AtomicInteger(0));
+      }
+      int concurrentQueryNum = _concurrentQueriesCountMap.get(serverHost).get();
+      if (concurrentQueryNum > _config.getMaxBacklogPerServer()) {
+        throw new PinotException(ErrorCode.PINOT_QUERY_BACKLOG_FULL,
+            "Reaching server query max backlog size is - " + _config.getMaxBacklogPerServer());
+      }
+      _concurrentQueriesCountMap.get(serverHost).incrementAndGet();
+      AsyncQueryResponse asyncQueryResponse;
+      QueryRouter nextAvailableQueryRouter = getNextAvailableQueryRouter();
+      if (TableNameBuilder.getTableTypeFromTableName(brokerRequest.getQuerySource().getTableName())
+          == TableType.REALTIME) {
+        asyncQueryResponse = nextAvailableQueryRouter.submitQuery(requestId, rawTableName, null, null, brokerRequest,
+            routingTable, connectionTimeoutInMillis);
+      } else {
+        asyncQueryResponse = nextAvailableQueryRouter
+            .submitQuery(requestId, rawTableName, brokerRequest, routingTable, null, null, connectionTimeoutInMillis);
+      }
+      Map<ServerInstance, DataTable> serverInstanceDataTableMap = gatherServerResponses(
+          ignoreEmptyResponses,
+          routingTable,
+          asyncQueryResponse,
+          brokerRequest.getQuerySource().getTableName());
+      _queryRouters.offer(nextAvailableQueryRouter);
+      _concurrentQueriesCountMap.get(serverHost).decrementAndGet();
+      return serverInstanceDataTableMap;
+    });
+    return serverResponseMap;
+  }
+
+  private QueryRouter getNextAvailableQueryRouter() {
+    QueryRouter queryRouter = _queryRouters.poll();
+    while (queryRouter == null) {
+      try {
+        Thread.sleep(200L);
+      } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)